packages/teleterm/src/ui/services/workspacesService/accessRequestsService.ts

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import { ResourceKind } from 'e-teleterm/ui/DocumentAccessRequests/NewRequest/useNewRequest';
+import { AccessRequest } from 'e-teleport/services/workflow';
+
+import { PendingAccessRequest } from '../workspacesService';
+
+export class AccessRequestsService {
+  constructor(
+    private getState: () => {
+      isAccessRequestsBarCollapsed: boolean;
+      pendingAccessRequest: PendingAccessRequest;
+      assumed: Record<string, AccessRequest>;
+    },
+    private setState: (
+      draftState: (draft: {
+        isAccessRequestsBarCollapsed: boolean;
+        pendingAccessRequest: PendingAccessRequest;
+        assumed: Record<string, AccessRequest>;
+      }) => void
+    ) => void
+  ) {}
+
+  getCollapsed() {
+    return this.getState().isAccessRequestsBarCollapsed;
+  }
+
+  toggleBar() {
+    this.setState(draftState => {
+      draftState.isAccessRequestsBarCollapsed =
+        !draftState.isAccessRequestsBarCollapsed;
+    });
+  }
+
+  getPendingAccessRequest() {
+    return this.getState()?.pendingAccessRequest;
+  }
+
+  getAssumed() {
+    return this.getState().assumed;
+  }
+
+  getAssumedRoles() {
+    const assumed = this.getState().assumed;
+    let roles = [];
+    Object.keys(assumed).forEach(key => {
+      const request = assumed[key];
+      const newRoles = request.roles.filter(r => !roles.includes(r));
+      roles = [...roles, ...newRoles];
+    });
+
+    return roles;
```

Comment:
I believe this would always return an empty array because the `list` in this case is the array being iterated over, not the one being built. I do like the flatmap part. I think this would work well
```js
    return [
      ...new Set(
        Object.values(this.getAssumed()).flatMap(request => request.roles)
      ),
    ];
```